internals/workloads/workloads.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/canonical/pebble/internals/plan"
+)
+
+type Workload struct {
+	// Basic details
+	Name     string        `yaml:"-"`
+	Override plan.Override `yaml:"override,omitempty"`
+
+	// Options for command execution
+	Environment map[string]string `yaml:"environment,omitempty"`
+	UserID      *int              `yaml:"user-id,omitempty"`
+	User        string            `yaml:"user,omitempty"`
+	GroupID     *int              `yaml:"group-id,omitempty"`
+	Group       string            `yaml:"group,omitempty"`
+}
+
+func (w *Workload) validate() error {
+	if w.Name == "" {
+		return errors.New("cannot have an empty name")
+	}
+	// Value of Override is checked in the (*WorkloadSection).combine() method
+	return nil
+}
+
+func (w *Workload) copy() *Workload {
+	copied := *w
+	copied.Environment = maps.Clone(w.Environment)
+	copied.UserID = copyPtr(w.UserID)
+	copied.GroupID = copyPtr(w.GroupID)
+	return &copied
+}
+
+func (w *Workload) merge(other *Workload) {
+	if len(other.Environment) > 0 {
+		w.Environment = makeMapIfNil(w.Environment)
+		maps.Copy(w.Environment, other.Environment)
+	}
+	if other.UserID != nil {
+		w.UserID = copyPtr(other.UserID)
+	}
+	if other.User != "" {
+		w.User = other.User
+	}
+	if other.GroupID != nil {
+		w.GroupID = copyPtr(other.GroupID)
+	}
+	if other.Group != "" {
+		w.Group = other.Group
+	}
+}
+
+func (w *Workload) Equal(other *Workload) bool {
+	return reflect.DeepEqual(w, other)
+}
+
+const WorkloadsField = "workloads"
+
+var (
+	_ plan.Section          = (*Workloads)(nil)
+	_ plan.SectionExtension = (*Workloads)(nil)
+)
+
+type Workloads struct {
+	Entries map[string]*Workload `yaml:",inline"`
+}
+
+func (w *Workloads) IsZero() bool {
+	return len(w.Entries) == 0
+}
+
+func (w *Workloads) Validate() error {
+	for name, workload := range w.Entries {
+		if workload == nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q: cannot have a null value", name),
+			}
+		}
+		if err := workload.validate(); err != nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q: %v", name, err),
+			}
+		}
+	}
+	return nil
+}
+
+func (w *Workloads) combine(other *Workloads) error {
+	for name, workload := range other.Entries {
+		w.Entries = makeMapIfNil(w.Entries)
+		switch workload.Override {
+		case plan.MergeOverride:
+			if current, ok := w.Entries[name]; ok {
+				copied := current.copy()
+				copied.merge(workload)
+				w.Entries[name] = copied
+				break
+			}
+			fallthrough
+		case plan.ReplaceOverride:
+			w.Entries[name] = workload.copy()
+		case plan.UnknownOverride:
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`workload %q: must define an "override" policy`, name),
+			}
+		default:
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`workload %q: has an invalid "override" policy: %q`, name, workload.Override),
+			}
+		}
+	}
+	return nil
+}
+
+func (*Workloads) CombineSections(sections ...plan.Section) (plan.Section, error) {
```

Comment:
This change where ```Workloads``` is now the extension type, and also creates ```Workloads``` seems a bit strange.